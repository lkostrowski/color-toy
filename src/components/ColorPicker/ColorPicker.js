import React from 'react';
import PropTypes from 'prop-types';
import ColorList from '../ColorList/ColorList';
import Input from '../Input/Input';
import Button from '../Button/Button';
import { IColor } from '../../models/Color.interface';
import { Headline, inputStyleOverrides, listStylesOverrides, Wrapper } from './ColorPicker.ui';

/**
 * Color picker component, containing headline, search input and colors list with button
 */
const ColorPicker = ({
    colors, onAccept, onColorSelected, selectedColor, onInputChange,
}) => (
    <Wrapper>
        <Headline>
            Pick your fancy color!
        </Headline>
        <Input
            className={inputStyleOverrides}
            onChange={onInputChange}
        />
        <ColorList
            className={listStylesOverrides}
            items={colors}
            onColorClicked={onColorSelected}
            selectedColor={selectedColor}
        />
        <Button onClick={onAccept}>Accept</Button>
    </Wrapper>
);

ColorPicker.propTypes = {
    colors: PropTypes.arrayOf(PropTypes.shape({})),
    onAccept: PropTypes.func.isRequired,
    onColorSelected: PropTypes.func.isRequired,
    selectedColor: IColor,
    onInputChange: PropTypes.func.isRequired,
};

export default ColorPicker;
