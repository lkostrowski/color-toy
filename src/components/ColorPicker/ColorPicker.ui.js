import { rgba } from 'polished';
import glamorous from 'glamorous/';
import colors from '../../colors';
import { css } from 'glamor';

export const Wrapper = glamorous.div({
    background: `linear-gradient(to bottom, ${colors.light}, ${rgba(colors.light, 0.6)})`,
    padding: '1rem',
    height: '100%',
    minHeight: '40rem',
    display: 'flex',
    flexDirection: 'column',
    boxShadow: `3px 3px 20px 3px ${rgba(colors.darkGray, 0.2)}`,
});

export const Headline = glamorous.h1({
    fontWeight: '300',
    color: colors.midGray,
    textTransform: 'uppercase',
    textAlign: 'center',
    letterSpacing: '0.05em',
    marginBottom: '2rem',
});

export const inputStyleOverrides = css({
    textTransform: 'uppercase',
});

export const listStylesOverrides = css({
    margin: '2rem 0 auto 0',
});
