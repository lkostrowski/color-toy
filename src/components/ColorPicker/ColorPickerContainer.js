import React, {Component} from 'react'
import PropTypes from 'prop-types'
import ColorPicker from './ColorPicker';
import {connect} from "react-redux";
import {IColor} from "../../models/Color.interface";
import {getVisibleColors} from "../../models/getSuggestedColors";

class ColorPickerContainer extends Component {
    static defaultProps = {};
    static propTypes = {
        colors: PropTypes.arrayOf(IColor),
        updateSearchQuery: PropTypes.func.isRequired,
        setActiveColor: PropTypes.func.isRequired,
    };

    state = {
        selectedColor: {}
    };

    /**
     * Autobind decorator could be useful
     */
    constructor() {
        super();
        this.onColorSelected = this.onColorSelected.bind(this);
        this.onInputChange = this.onInputChange.bind(this);
        this.onAcceptClick = this.onAcceptClick.bind(this);
    }

    onAcceptClick() {
        if (!this.state.selectedColor) return;
        if (this.props.activeColor.name === this.state.selectedColor.name) return;

        this.props.setActiveColor(this.state.selectedColor);
    }

    onColorSelected(color) {
        this.setState({
            selectedColor: color
        })
    }

    onInputChange(e) {
        const query = e.target.value;
        this.props.updateSearchQuery(query);
    }

    render() {
        const {colors} = this.props;
        const {selectedColor} = this.state;

        return (
            <ColorPicker onColorSelected={this.onColorSelected}
                         colors={colors}
                         selectedColor={selectedColor}
                         onInputChange={this.onInputChange}
                         onAccept={this.onAcceptClick}
            />
        );
    }
}

const mapState = state => ({
    colors: getVisibleColors(state),
    activeColor: state.activeColor
});

const mapDispatch = dispatch => ({
    setActiveColor: color => dispatch.activeColor.setActiveColor(color),
    updateSearchQuery: query => dispatch.autosuggest.updateSearchQuery(query)
});

export default connect(mapState, mapDispatch)(ColorPickerContainer);
