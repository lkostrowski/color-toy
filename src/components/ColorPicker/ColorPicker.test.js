import ColorPicker from './ColorPicker';
import React from 'react';
import renderer from 'react-test-renderer';

describe('ColorPicker component', () => {
    it('render expected html', () => {
        const noop = () => {
        };
        const testColors = [
            {
                name: 'test-color1',
                hex: '#ffffff'
            },
            {
                name: 'test-color2',
                hex: '#000000'
            }
        ];
        const testActiveColor = testColors[0];


        const markup = renderer.create(
            <ColorPicker onColorSelected={noop} onAccept={noop} onInputChange={noop}
                         selectedColor={testActiveColor} colors={testColors}/>
        ).toJSON();

        expect(markup).toMatchSnapshot();
    });
});
