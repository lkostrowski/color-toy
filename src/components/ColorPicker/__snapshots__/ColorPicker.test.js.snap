// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker component render expected html 1`] = `
<div
  className="css-1ctrea"
>
  <h1
    className="css-1kkvtqk"
  >
    Pick your fancy color!
  </h1>
  <input
    className="css-1pwtocw"
    onChange={[Function]}
  />
  <ul
    className="css-1s18ngn"
  >
    <li
      className="css-1oxu79"
      onClick={[Function]}
    >
      <div
        className="css-x2kxff"
      />
      <span
        className="css-nil"
      >
        test-color1
      </span>
    </li>
    <li
      className="css-1x7obq3"
      onClick={[Function]}
    >
      <div
        className="css-gbdd7j"
      />
      <span
        className="css-nil"
      >
        test-color2
      </span>
    </li>
  </ul>
  <button
    className="css-1kigtsu"
    onClick={[Function]}
  >
    Accept
  </button>
</div>
`;
